ambry-clustermap/src/main/java/com.github.ambry.clustermap/DynamicClusterChangeHandler.java

```diff
@@ -0,0 +1,500 @@
+    logger.info("Updating replicas info for existing node {}", instanceName);
+    List<String> sealedReplicas = getSealedReplicas(instanceConfig);
+    List<String> stoppedReplicas = getStoppedReplicas(instanceConfig);
+    AmbryDataNode dataNode = instanceNameToAmbryDataNode.get(instanceName);
+    ConcurrentHashMap<String, AmbryReplica> currentReplicasOnNode = ambryDataNodeToAmbryReplicas.get(dataNode);
+    ConcurrentHashMap<String, AmbryReplica> replicasFromConfig = new ConcurrentHashMap<>();
+    Map<String, Map<String, String>> diskInfos = instanceConfig.getRecord().getMapFields();
+    for (Map.Entry<String, Map<String, String>> entry : diskInfos.entrySet()) {
+      String mountPath = entry.getKey();
+      Map<String, String> diskInfo = entry.getValue();
+      Optional<AmbryDisk> potentialDisk =
+          ambryDataNodeToAmbryDisks.get(dataNode).stream().filter(d -> d.getMountPath().equals(mountPath)).findFirst();
+      AmbryDisk disk = potentialDisk.orElse(null);
+      if (disk == null) {
+        logger.info("Temporarily don't support adding new disk to existing node.");
+        // TODO support dynamically adding disk in the future
+        continue;
+      }
+      String replicasStr = diskInfo.get(ClusterMapUtils.REPLICAS_STR);
+      if (!replicasStr.isEmpty()) {
+        for (String replicaInfo : replicasStr.split(ClusterMapUtils.REPLICAS_DELIM_STR)) {
+          String[] info = replicaInfo.split(ClusterMapUtils.REPLICAS_STR_SEPARATOR);
+          // partition name and replica name are the same.
+          String partitionName = info[0];
+          if (currentReplicasOnNode.containsKey(partitionName)) {
+            // if replica is already present
+            AmbryReplica existingReplica = currentReplicasOnNode.get(partitionName);
+            // 1. directly add it into "replicasFromConfig" map
+            replicasFromConfig.put(partitionName, existingReplica);
+            // 2. update replica seal/stop state
+            updateReplicaStateAndOverrideIfNeeded(existingReplica, sealedReplicas, stoppedReplicas);
+          } else {
+            // if this is a new replica and doesn't exist on node
+            logger.info("Adding new replica {} to existing node {}", partitionName, instanceName);
+            long replicaCapacity = Long.valueOf(info[1]);
+            String partitionClass = info.length > 2 ? info[2] : clusterMapConfig.clusterMapDefaultPartitionClass;
+            // this can be a brand new partition that is added to an existing node
+            AmbryPartition mappedPartition =
+                new AmbryPartition(Long.valueOf(partitionName), partitionClass, helixClusterManagerCallback);
+            // Ensure only one AmbryPartition instance exists for specific partition.
+            mappedPartition = clusterChangeHandlerCallback.addPartitionIfAbsent(mappedPartition, replicaCapacity);
+            ensurePartitionAbsenceOnNodeAndValidateCapacity(mappedPartition, dataNode, replicaCapacity);
```

Comment:
General question: What happens when exceptions get thrown inside of helix listener callbacks? Would future notifications be received?

Throwing at this point in the middle of the loop could create some inconsistency since clusterChangeHandlerCallback.addReplicasToPartition is called on every iteration of the loop, but ambryDataNodeToAmbryReplicas is updated after the loop. If this is an issue, you could move the clusterChangeHandlerCallback.addReplicasToPartition call to after the loop.

Reply:
I need to check with Helix team. Yeah, we need to move `addReplicasToPartition` out of loop.